libopenage/datastructure/concurrent_queue.h

```diff
@@ -47,25 +46,46 @@ class ConcurrentQueue {
 		return this->queue.front();
 	}
 
-	/** Removes the front item of the queue and returns it. */
-	T &pop() {
+	/** Copies the front item in the queue and removes it from the queue. */
+	template<typename... None, typename U = T>
+	T& pop([[maybe_unused]] typename std::enable_if_t<!std::is_move_constructible_v<U> and std::is_copy_constructible_v<U>>* t = NULL) {
+		static_assert(sizeof...(None) == 0, "User-specified template arguments are prohibited.");
 		std::unique_lock<mutex_t> lock{this->mutex};
-		while (this->queue.empty()) {
-			this->elements_available.wait(lock);
-		}
-		auto &item = this->queue.front();
+		auto &item = front();
 		this->queue.pop();
 		return item;
```

Comment:
this is a memory error, we have to return a `T`, and not a `T&`. so `auto item = this->front();` and `return item;`